tiny_cnn/core/backend_avx.h

```diff
@@ -0,0 +1,285 @@
+
+        std::fill(a.begin(), a.end(), float_t(0));
+
+        kernels::avx_conv2d_kernel(*params_c_,
+            in, W, bias, a, layer_->get_parallelize());
+    }
+
+    void conv2d(cnn_size_t                 index,
+                const std::vector<vec_t*>& in_data,
+                const std::vector<vec_t*>& out_data,
+                std::vector<vec_t*>&       out_grad,
+                std::vector<vec_t*>&       in_grad) {
+        conv_layer_worker_specific_storage& cws =
+            (*conv_layer_worker_storage_)[index];
+
+        const vec_t& prev_out = *(cws.prev_out_padded_);
+        const vec_t& W  = *in_data[1];
+        vec_t&       dW = *in_grad[1];
+        vec_t&       db = *in_grad[2];
+        vec_t&       curr_delta = *out_grad[1];
+        vec_t*       prev_delta = (params_c_->pad_type == padding::same) ?
+                                   &cws.prev_delta_padded_ : in_grad[0];
+
+        assert(W.size() == params_c_->weight.size());
+        assert(dW.size() == params_c_->weight.size());
+        assert(curr_delta.size() ==  layer_->out_shape()[0].size());
+
+        backward_activation(*out_grad[0], *out_data[0], curr_delta);
+
+        std::fill(prev_delta->begin(), prev_delta->end(), float_t(0));
+
+        kernels::avx_conv2d_back_kernel(*params_c_,
+            prev_out, W, dW, db, curr_delta, prev_delta);
+
+        if (params_c_->pad_type == padding::same) {
+            copy_and_unpad_delta(cws.prev_delta_padded_, *in_grad[0]);
+        }
+    }
+
+    void deconv2d(cnn_size_t        index,
+        const std::vector<vec_t*>&  in_data,
+        std::vector<vec_t*>&        out_data) {
+        (*deconv_layer_worker_storage_)[index].prev_out_ = in_data[0];
+        const vec_t& W   = *in_data[1];
+        const vec_t& bias = *in_data[2];
+        vec_t&       a   = *out_data[1];
+        const vec_t &in  = *in_data[0]; // input
+
+        std::fill(a.begin(), a.end(), float_t(0));
+
+        kernels::avx_deconv2d_kernel(*params_d_,
+            in, W, bias, a, layer_->get_parallelize());
+
+        copy_and_unpad_output(a, static_cast<int>(index));
+    }
+
+    void deconv2d(cnn_size_t                 index,
+                       const std::vector<vec_t*>& in_data,
+                       const std::vector<vec_t*>& out_data,
+                       std::vector<vec_t*>&       out_grad,
+                       std::vector<vec_t*>&       in_grad) {
+
+        deconv_layer_worker_specific_storage& cws =
+            (*deconv_layer_worker_storage_)[index];
+        if (params_d_->pad_type == padding::same)
+            copy_and_pad_delta(cws.curr_delta_padded, *in_grad[0]);
+
+        const vec_t& prev_out = *(cws.prev_out_);
+        const vec_t& W = *in_data[1];
+        vec_t&       dW = *in_grad[1];
+        vec_t&       db = *in_grad[2];
+        vec_t&       curr_delta = (params_d_->pad_type == padding::same) ? cws.curr_delta_padded : *out_grad[1];
+        vec_t*       prev_delta = in_grad[0];
```

Comment:
How should I format the code? Is it very different from `tiny_cnn/core/backend_tiny.h`